Support for DPS5015 and CC+CV mode (CurrentLimit)

cleverfox opened the pull request:
Hello. I made support for DPS5015. And now it almost usable. 
Also CC + CV mode (as in original DPS firmware) which is userful in many cases. I didn't made changes for binary protocol support, but I made full support in text protocol. As for now there are few new commands:
cc - CC mode
cv - CV mode
cl - Current Limit mode (CV unless current less than setting, else CC mode)
u - set voltage (instead of v command)
i - set current
I changed stat, and now it showing complete state.
Also I moved coefficients and offsets for ADC and DAC (only for current yet) into dps-model.h to support different shunt. And described formulas to calculate factor and offset;
Also I added software debounce for power enable button.
I need your help with mode image, I bithacked CV image, to make a little bit different image, because I have no idea how to cook this images.  
As for now I have not fixed LCD hanging yet. But as temporary hack I make reinitialization of LCD in "on" CLI command. I will try to fix is this problem.

felixekman commented:
Do you know if these changed are compatible with the DPH3205? I've been thinking about adding support for it, but I can't really wrap my head around it. How did you go about to add support for the DPS5015, schematic and test?

toglader commented:
BTW, Why have you changed all (de facto standard) \n to \r\n (and in some cases to \r)?

cleverfox commented:
All terminal apps, which I ever used do not do carrier return on \n. When I writing unix apps, I always using \n, but for serail ports it is not working for me, and I using \r\n. I not happy with it, but without \r it is looks unusable :)

cleverfox commented:
I have no idea about DPH3205, I have no such device. But you can try :)

toglader commented:
Usually terminal programs has configurable line endings... You should find better terminal emulator - for example putty :)

kanflo commented on opendps/cli.c:
```diff
@@ -83,9 +83,9 @@ void cli_run(const cli_command_t cmds[], const uint32_t num, char *line)
         if (cmds[i].cmd && strcmp(argv[0], cmds[i].cmd) == 0) {
             if (cmds[i].min_arg > argc-1 || cmds[i].max_arg < argc-1) {
-                printf("Wrong number of arguments:\n");
+                printf("Wrong number of arguments:\r\n");
```
I agree with @toglader the line endings should not be changed.

cleverfox replied:
I agree. I trying to find a way to work with \n EOLs.

kanflo commented on opendps/dps-model.h:
```diff
@@ -3,7 +3,41 @@
 // if no other DPSxxxx model is specified, we will assume DPS5005
 
-// #define DPS5015 
+#define DPS5015 
```
This would surprise current 5005 users. I could go with the makefile complaining unless the model was specified at compile time.

cleverfox replied:
Ok, I forgot to change back it

kanflo commented on opendps/dps-model.h:
```diff
@@ -3,7 +3,41 @@
+#define DPS5015 
 
+/*
+ * K - angle factor
```
Thanks! This is a good idea.

kanflo commented on opendps/hw.c:
```diff
@@ -704,12 +706,16 @@ void BUTTON_M2_isr(void)
 void BUTTON_ENABLE_isr(void)
 {
     static bool falling = true;
+    static uint32_t lastclick=0;
```
Great addon. Still wondering why this is not needed for the 5005s...

kanflo commented on opendps/ui.c:
```diff
@@ -386,48 +480,22 @@ void ui_update_values(uint32_t v_in, uint32_t v_out, uint32_t i_out)
-            tft_putch(1, 'V', voltage_positions[5], YPOS_VOLTAGE, font_1_widths[font_1_num_glyphs-1], h, false);
-        }
-
+        display_voltage(temp);
```
Glad to have this cleaned up, thanks!

kanflo commented on opendps/command_handler.c:
```diff
@@ -107,20 +141,27 @@ static void on_cmd(uint32_t argc, char *argv[])
     ui_update_power_status(true);
+
+
+#warning TEMPORARY HACK TO UNFREEZE LCD
```
This should go with an ifdef UNFREEZE_LDC_HACK coming from dps-model.h and of course need to be set for the 5005s.

cleverfox replied:
Ok. I agree. I will fix

kanflo commented on opendps/command_handler.c:
```diff
@@ -130,15 +171,66 @@ static void stat_cmd(uint32_t argc, char *argv[])
-    printf(" V_in  : %2lu.%02lu V\n", v_in/1000,  (v_in%1000)/10);
-    printf(" V_out : %2lu.%02lu V (%s)\n", v_out/1000, (v_out%1000)/10, pwrctl_vout_enabled() ? "enabled" : "disabled");
-    printf(" I_out : %2lu.%03lu A\n", i_out/1000, i_out%1000);
+    power_mode_t mode=ui_get_power_mode();
```
If I may be picky, I'd prefer the already set style is followed. I tend to adhere to this at work, even when I don't like it ;) Eg. ```mode = ui_get_power_mode();```

cleverfox replied:
Sorry, my english is not so good. I really don't understand what's wrong here.

cleverfox commented:
Now I believe that I fixed LCD freeze for DPS5015. 
Rolled back \r\n to \n. I found way to use \n EOLs in minicom. 
Fixed default device to DPS5005. 
Can anybody help me with CL image (like CC and CV at the bottom of screen)?

cleverfox commented:
Also I found a bug which occur only sometimes. When I push to encoder selected value is decremented by 1. Can anybody confirm this bug?

kanflo commented:
@cleverfox I can fix the CL image for you. I have never observed the encoder bug though.

cleverfox commented:
@kanflo Thank you. I need .h file for "cl" image. As for now I not commited it yet, but it used in sources.

cleverfox commented:
Hello, guys. I made calibration voltage ADC and DACs for my DPS5015. Also I added example of calibration calculations inside dps-model.c for reference.

roger- commented:
Question: is configuring this from the device UI supported or do you need to send commands over serial?

vogelfreiheit commented:
I might send an e-mail to cleverfox since I encountered issues with the master branch for the 2.3 PCB version of the 5015. Will be testing the cc branch now. How much of this is merged into master or expected to be?

kanflo commented:
Everything from the cc branch is scheduled for merge to master. Obviously that did not happen last month so I am hoping for December. I need to get my AAduino Zero crowd funding project launched first.

Out of curiosity, how do you identify your DPS5015 as "2.3"? I have not seen any silk screen markings on my '5005s.

kanflo commented:
@roger- I believe this is only supported through the text protocol, if memory serves me well...

vogelfreiheit commented:
I finally got to do some testing and work around this. I recalibrated for the latest PCB... managed to get the power output within spec/precision. However, I think we should have a self-calibration facility of some sort. A debug screen would also be great, possibly with simple text rendering.... no need for fancy fonts. DAC, ADC, etc and some registers printed to screen, since it is tedious to rely on BOTH UART and JTAG for this, and actually outputting to GDB or console from the firmware is not trivial if using interrupt hacks. I did my calibration with some gdb scripting to pull values at certain breakpoints and then continue execution.
